File: src/main/java/core/reader/Reader.java
package core.reader;

import core.exceptions.IllegalSyntaxException;
import core.reader.parsers.Result;
import core.scm.SCMBoolean;
import core.scm.SCMCons;
import core.scm.SCMSymbol;
import core.scm.SCMVector;
import core.scm.specialforms.Quasiquote;
import core.scm.specialforms.Quote;
import core.scm.specialforms.Unquote;
import core.scm.specialforms.UnquoteSplicing;
import core.utils.NumberUtils;

import java.io.*;
import java.text.ParseException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static core.utils.NumberUtils.*;

public class Reader implements IReader {

  private static final SCMSymbol DOT = new SCMSymbol(".");
  private static final String LINE_BREAKS = "\n\f\r";
  private static final String WHITESPACES = (char)0x0B + " \t" + LINE_BREAKS;
  // <delimiter> --> <whitespace> | ( | ) | " | ;
  private static final String DELIMITERS = WHITESPACES + "()\";" + '\u0000' + '\uffff';

  private static final Map<String, Character> NAMED_CHARS = new HashMap<>();
  static {
    // TODO Platform-dependent line separator?
    NAMED_CHARS.put("newline",   '\n');
    NAMED_CHARS.put("linefeed",  '\n');
    NAMED_CHARS.put("space",     ' ');
    NAMED_CHARS.put("tab",       '\t');
    NAMED_CHARS.put("return",    '\r');
    NAMED_CHARS.put("backspace", '\b');
    NAMED_CHARS.put("page",      '\f');
    NAMED_CHARS.put("alarm",     '\u0007');
    NAMED_CHARS.put("vtab",      '\u000B');
    NAMED_CHARS.put("esc",       '\u001B');
    NAMED_CHARS.put("escape",    '\u001B');
    NAMED_CHARS.put("delete",    '\u007F');
    NAMED_CHARS.put("null",      Character.MIN_VALUE);
    NAMED_CHARS.put("nul",       Character.MIN_VALUE);
  }

  private static final Map<Character, String> CODEPOINTS = new HashMap<>();
  static {
    for (Map.Entry<String, Character> entry : NAMED_CHARS.entrySet()) {
      CODEPOINTS.put(entry.getValue(), entry.getKey());
    }
  }

  public static String charToNamedChar(Character ch) {
    return CODEPOINTS.get(ch);
  }

  private static boolean isValid(int i) {
    return i > -1 && i < 65535;
  }

  private static boolean isExactness(char c) {
    return c == 'i' || c == 'e';
  }

  private static boolean isRadix(char c) {
    return "bodxBODX".indexOf(c) > -1;
  }

  @Override
  public Object read(String string) {
    PushbackReader reader = new PushbackReader(new StringReader(string), 2);
    try {
      Object token = nextToken(reader);
      if (DOT.equals(token)) {
        throw new IllegalSyntaxException("Illegal use of '.'");
      }
      return token;
    } catch (IOException e) {
      e.printStackTrace();
    } catch (ParseException e) {
      e.printStackTrace();
    }
    return null;
  }

  @Override
  public Object read(InputStream inputStream) {
    PushbackReader reader = new PushbackReader(new BufferedReader(new InputStreamReader(inputStream)), 2);
    try {
      Object token = nextNonNullToken(reader);
      if (DOT.equals(token)) {
        throw new IllegalSyntaxException("Illegal use of '.'");
      }
      return token;
    } catch (IOException e) {
      e.printStackTrace();
    } catch (ParseException e) {
      e.printStackTrace();
    }
    return null;
  }

  /* Skip all null tokens and return the first non-null */
  private static Object nextNonNullToken(PushbackReader reader) throws IOException, ParseException {
    Object token;
    while ((token = nextToken(reader)) == null) {/* Read */};
    return token;
  }

  /**
   * Read next token
   */
  private static Object nextToken(PushbackReader reader) throws IOException, ParseException {
    int i;
    if (!isValid(i = reader.read())) {
      return null;
    }
    char c = (char)i;
    switch (c) {
      case '\'':
        return readQuote(reader, Quote.QUOTE.symbol());
      case '`':
        return readQuote(reader, Quasiquote.QUASIQUOTE.symbol());
      case ',': {
        char next = (char) reader.read();
        if (next == '@') {
          return readQuote(reader, UnquoteSplicing.UNQUOTE_SPLICING.symbol());
        } else {
          reader.unread(next);
          return readQuote(reader, Unquote.UNQUOTE.symbol());
        }
      }
      case '#':
        char next = (char) reader.read();
        if (next == '(') {
          return readVector(reader);
        } else {
          reader.unread(next);
          reader.unread(c);
          return readAtom(reader);
        }
      case '(':
        return readList(reader);
      case ')':
        throw new IllegalSyntaxException("Unexpected list terminator: ')'");
      default: {
        if (Character.isWhitespace(c)) {
          while (isValid(c = (char)reader.read()) && Character.isWhitespace(c)) {
            /* Skip whitespaces */
          }
          reader.unread(c);
          return null;
        } else {
          reader.unread(c);
          return readAtom(reader);
        }
      }
    }
  }

  /**
   * Read atom
   */
  private static Object readAtom(PushbackReader reader) throws IOException, ParseException {
    char c = (char)reader.read();
    char next = (char)reader.read();
    reader.unread(next);
    /* Decimal number */
    if (c != '#' && isValidForRadix(c, 10)) {
      // dot?
      if (c == '.' && DELIMITERS.indexOf(next) > -1) {
        return DOT;
      }
      reader.unread(c);
      /* Read identifier, not a number */
      String number = readIdentifier(reader).toString();
      if (NumberUtils.SPECIAL_NUMBERS.containsKey(number)) {
        return NumberUtils.SPECIAL_NUMBERS.get(number);
      }
      /* Now check if it IS a valid number */
      return preProcessNumber(number, 'e', 10);
    } else if (c == ';') {
      return readComment(reader);
    } else if (c == '"') {
      return readString(reader);
    } else if (c == '#') {
      Object result = readHash(reader);
      if (result != null) {
        return result;
      }
    } else {
      reader.unread(c);
      return readIdentifier(reader);
    }
    throw new IllegalSyntaxException("Unknown atom!");
  }

  private static Object readHash(PushbackReader reader) throws ParseException, IOException {
    char next = (char)reader.read();
    if (next == '\\') {
      return readCharacter(reader);
    } else if (next == 't' || next == 'T') {
      return SCMBoolean.TRUE;
    } else if (next == 'f' || next == 'F') {
      return SCMBoolean.FALSE;
    } else if (isRadix(next) || isExactness(next)) {
      reader.unread(next);
      reader.unread('#');

      /* Read identifier, not a number */
      String number = readIdentifier(reader).toString();

      /* Read radix and/or exactness and a number */
      Character radixChar = null;
      Character exactness = null;
      Result parse = EXACTNESS_RADIX.parse(number);
      if (parse.getType() == Result.Type.SUCCESS) {
        List<String> match = parse.getMatch();
        exactness = match.get(0).charAt(1);
        if (match.size() > 1) {
          radixChar = match.get(1).charAt(1);
        }
      } else {
        parse = RADIX_EXACTNESS.parse(number);
        if (parse.getType() == Result.Type.SUCCESS) {
          List<String> match = parse.getMatch();
          radixChar = match.get(0).charAt(1);
          if (match.size() > 1) {
            exactness = match.get(1).charAt(1);
          }
        }
      }
      exactness = (exactness == null) ? 'e' : exactness;
      radixChar = (radixChar == null) ? 'd' : radixChar;

      number = parse.getRest();
      if (number.isEmpty()) {
        throw new IllegalSyntaxException("Bad number: no digits!");
      }

      /* Check if we got exactness or radix */
      return preProcessNumber(number, exactness, getRadixByChar(radixChar));
    }
    return null;
  }

  /**
   * Read a quoted form abbreviation
   *
   * Syntax:
   * <quote> -> '<form>
   * <quasiquote> -> `<form>
   * <unquote> -> ,<form>
   * <unquote-splicing> -> ,@<form>
   */
  private static Object readQuote(PushbackReader reader, SCMSymbol symbol) throws ParseException, IOException {
    List<Object> quote = SCMCons.list(symbol);
    quote.add(nextNonNullToken(reader));
    return quote;
  }

  /**
   * Read identifier
   *
   * Syntax:
   * <identifier> --> <initial> <subsequent>* | <peculiar identifier>
   */
  private static Object readIdentifier(PushbackReader reader) throws IOException {
    StringBuilder identifier = new StringBuilder();
    int i = reader.read();
    char c = (char)i;
    while (isValid(i) && DELIMITERS.indexOf(c) < 0) {
      identifier.append(c);
      i = reader.read();
      c = (char)i;
    }
    reader.unread(c);
    return new SCMSymbol(identifier.toString());
  }

  /**
   * Read a comment
   *
   * Syntax:
   * <comment> --> ;  <all subsequent characters up to a line break>
   */
  private static String readComment(PushbackReader reader) throws IOException {
    int i;
    while (isValid(i = reader.read()) && (LINE_BREAKS.indexOf((char)i) < 0)) {
      /* Read everything until line break */
    }
    /* Comments are ignored, so just return null */
    return null;
  }

  /**
   * Read a Number
   *
   * Syntax:
   * (See above for full syntax)
   */
  private static String readNumber(PushbackReader reader, int radix) throws ParseException, IOException {
    StringBuilder number = new StringBuilder();
    int i = reader.read();
    char c = (char)i;
    if (c == '.') {
      number.append('0');
    }
    while (isValid(i) && (isValidForRadix(c, radix))) {
      number.append(c);
      i = reader.read();
      c = (char)i;
    }
    /* Ensure we have 1 sign/dot max */
    if ((number.indexOf("+") != number.lastIndexOf("+")) ||
        (number.indexOf("-") != number.lastIndexOf("-")) ||
        (number.indexOf(".") != number.lastIndexOf("."))) {

      throw new IllegalSyntaxException("Bad number!");
    }
    reader.unread(c);
    if (number.length() > 0 && number.charAt(number.length() - 1) == '.') {
      number.append('0');
    }
    return number.toString();
  }

  /**
   * Read a String
   *
   * Syntax:
   * <string> --> " <string element>* "
   * <string element> --> <any character other than " or \> | \" | \\
   */
  private static String readString(PushbackReader reader) throws ParseException, IOException {
    StringBuilder string = new StringBuilder();
    int i;
    char c;
    while ((isValid(i = reader.read())) && ((c = (char)i) != '"')) {
      // escaping
      if (c == '\\') {
        i = reader.read();
        char next = (char)i;
        if (next == '>') {
          throw new IllegalSyntaxException("Warning: undefined escape sequence in string - probably forgot backslash: #\\>");
        } else if (next == '"') {
          string.append(c).append(next);
          continue;
        }
      }
      string.append(c);
    }
    return string.toString();
  }

  /**
   * Read a Character
   *
   * Syntax:
   * <character> --> #\ <any character> | #\ <character name>
   * <character name> --> space | newline
   */
  private static Character readCharacter(PushbackReader reader) throws ParseException, IOException {
    int i;
    /* Check if it is a codepoint */
    if (isValid(i = reader.read()) && (Character.isDigit((char)i) || ((char)i == 'x'))) {
      char radixChar = (((char)i == 'x') || ((char)i == 'X')) ? 'x' : 'd';
      if (radixChar != 'x') {
        reader.unread((char)i);
      }
      String codepoint = readNumber(reader, getRadixByChar(radixChar));
      int cp = -1;
      if (radixChar == 'd') {
        /* Decimal digits, not a codepoint */
        if (codepoint.length() == 1) {
          return codepoint.charAt(0);
        }
        cp = Integer.parseInt(codepoint, 10);
      } else if (radixChar == 'x') {
        if (codepoint.isEmpty()) {
          return 'x';
        }
        cp = Integer.parseInt(codepoint, 16);
      }
      if (!Character.isValidCodePoint(cp)) {
        throw new IllegalSyntaxException("Bad codepoint: " + cp);
      }
      return (char)cp;
    }
    reader.unread((char)i);

    StringBuilder character = new StringBuilder();
    char c = (char)reader.read();
    do {
      character.append(c);
    } while ((isValid(i = reader.read())) && (DELIMITERS.indexOf(c = (char)i) < 0));
    reader.unread((char)i);

    // <character name>
    if (character.length() > 1) {
      Character namedChar = NAMED_CHARS.get(character.toString());
      if (namedChar == null) {
        throw new IllegalSyntaxException("Error: unknown named character: \"" + character + "\"");
      }
      return namedChar;
    }
    return character.charAt(0);
  }

  /**
   * Read list
   *
   * Syntax:
   * <list> -> (<list_contents>)
   */
  private static SCMCons<Object> readList(PushbackReader reader) throws ParseException, IOException {
    SCMCons<Object> list = SCMCons.NIL;
    int i;
    char c;
    /* Position of a dot (if we have it) */
    int dotPos = -1;
    /* Current index */
    int pos = -1;
    while (isValid(i = reader.read()) && ((c = (char)i) != ')')) {
      reader.unread(c);
      Object token = nextToken(reader);
      if (token != null) {
        pos += 1;
        /* Have no elements in a result list yet */
        if (SCMCons.NIL.equals(list)) {
          /* Create empty list (can't modify NIL) */
          list = SCMCons.list();
        }
        /* Check if current token is a dot */
        if (DOT.equals(token)) {
          /* Remember the position */
          dotPos = pos;
        }
        /* Add token read */
        list.add(token);
      }
    }
    /* It was a proper list */
    if (dotPos < 0) {
      return list;
    }

    /* Process improper list */
    if (dotPos != list.size() - 2) {
      throw new IllegalSyntaxException("Error: bad dotted pair form: " + list);
    }
    /* Remove dot */
    list.remove(dotPos);
    /* Convert list into cons */
    Object last = list.get(list.size() - 1);
    Object beforeLast = list.get(list.size() - 2);
    SCMCons<Object> cons = SCMCons.cons(beforeLast, last);
      /* Cons backwars */
    // TODO Do not iterate the same list again?
    for (int n = list.size() - 3; n >= 0; n--) {
      cons = SCMCons.cons(list.get(n), cons);
    }
    return cons;
  }

  /**
   * Read vector
   *
   * Syntax:
   * <vector> -> </vector>#(<vector_contents>)
   */
  private static SCMVector readVector(PushbackReader reader) throws ParseException, IOException {
    return new SCMVector(readList(reader).toArray());
  }
}

File: src/main/java/core/reader/Reader.kt
package core.reader

import core.exceptions.IllegalSyntaxException
import core.procedures.AFn
import core.scm.Cons
import core.scm.Keyword
import core.scm.MutableVector
import core.scm.Symbol
import core.scm.specialforms.Quasiquote
import core.scm.specialforms.Quote
import core.scm.specialforms.Unquote
import core.scm.specialforms.UnquoteSplicing
import core.utils.Utils.getRadixByChar
import core.utils.Utils.isValidForRadix
import core.utils.Utils.preProcessNumber
import java.io.*
import java.util.*
import java.util.regex.Pattern
import kotlin.collections.HashMap

open class Reader : IReader {

    internal lateinit var reader: PushbackReader

    internal constructor()

    constructor(inputStream: InputStream) {
        this.reader = PushbackReader(BufferedReader(InputStreamReader(inputStream)), 1)
    }

    companion object {

        private val DOT = Symbol.intern(".")

        private val DEREF = Symbol.intern("deref")

        private val LINE_BREAKS = "\n\r"
        private val WHITESPACES = LINE_BREAKS + "\u000B \t"
        // <delimiter> --> <whitespace> | ( | ) | " | ;
        private val DELIMITERS = WHITESPACES + ":;(){}[],\"\u0000\uffff"

        /* Allowed escape sequences. See: https://docs.oracle.com/javase/tutorial/java/data/characters.html */
        private val ESCAPED = HashMap<Char, Char>()

        init {
            ESCAPED.put('t', '\t')
            ESCAPED.put('b', '\b')
            ESCAPED.put('n', '\n')
            ESCAPED.put('r', '\r')
            ESCAPED.put('"', '\"')
            ESCAPED.put('\\', '\\')
        }

        val NAMED_CHARS: MutableMap<String, Char> = HashMap()

        init {
            NAMED_CHARS.put("newline", '\n')
            NAMED_CHARS.put("space", ' ')
            NAMED_CHARS.put("tab", '\t')
            NAMED_CHARS.put("return", '\r')
            NAMED_CHARS.put("backspace", '\b')
            NAMED_CHARS.put("alarm", '\u0007')
            NAMED_CHARS.put("vtab", '\u000B')
            NAMED_CHARS.put("esc", '\u001B')
            NAMED_CHARS.put("escape", '\u001B')
            NAMED_CHARS.put("delete", '\u007F')
            NAMED_CHARS.put("null", Character.MIN_VALUE)
            NAMED_CHARS.put("nul", Character.MIN_VALUE)
        }

        private fun isValid(i: Int): Boolean {
            return i > Character.MIN_VALUE.toInt() && i < Character.MAX_VALUE.toInt()
        }

        private fun isLineBreak(c: Char): Boolean {
            return LINE_BREAKS.indexOf(c) > -1
        }

        @JvmStatic fun isRadix(c: Char): Boolean {
            return "bodxBODX".indexOf(c) > -1
        }

        @JvmStatic fun isExact(c: Char): Boolean {
            return c == 'e' || c == 'E'
        }

        @JvmStatic fun isInexact(c: Char): Boolean {
            return c == 'i' || c == 'I'
        }

        @JvmStatic fun isExactness(c: Char): Boolean {
            return isExact(c) || isInexact(c)
        }
    }

    override fun read(): List<Any> {
        val tokens = ArrayList<Any>()
        try {
            var token = nextToken()
            while (token != null || tokens.isEmpty()) {
                when {
                    token != null -> tokens.add(token)
                }
                token = nextToken()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return tokens
    }

    @Throws(IOException::class)
    private fun readUntilDelimiter(): String {
        val token = StringBuilder()
        var i = reader.read()
        while (isValid(i) && DELIMITERS.indexOf(i.toChar()) < 0) {
            token.append(i.toChar())
            i = reader.read()
        }
        reader.unread(i.toChar().toInt())
        return token.toString()
    }

    /* Skip all null tokens and return the first non-null */
    @Throws(IOException::class)
    private fun nextNonNullToken(): Any {
        var token = nextToken()
        while (token == null) {
            /* Read */
            token = nextToken()
        }
        return token
    }

    /**
     * Read next token
     */
    @Throws(IOException::class)
    fun nextToken(): Any? {
        val i = reader.read()
        if (!isValid(i)) {
            return null
        }
        var c = i.toChar()
        /* Skip whitespaces until line break */
        if (Character.isWhitespace(c)) {
            while (isValid(c.toInt()) && Character.isWhitespace(c) && !isLineBreak(c)) {
                c = reader.read().toChar()
            }
        }
        /* Check if there is anything to read */
        if (!isValid(c.toInt()) || isLineBreak(c)) {
            return null
        }
        /* Decimal number */
        if (c != '#' && isValidForRadix(c, 10)) {
            /* Read identifier, not a number */
            val number = c + readUntilDelimiter()
            /* Now check if it IS a valid number */
            return preProcessNumber(number, null, 10)
        }
        when (c) {
            '\'' -> return readQuote(c)
            '`' -> return readQuote(c)
            ',' -> return readQuote(c)
            '@' -> return readDeref()
            '#' -> return readHash()
            '(' -> return readList(true, ')')
            '{' -> return readHashmap()
            '[' -> return readVector(']')
            ';' -> return readComment()
            '"' -> return readString()
            ':' -> return readKeyword()
            ')' -> throw IllegalSyntaxException("read: unexpected list terminator: " + c)
            '}' -> throw IllegalSyntaxException("read: unexpected terminator: " + c)
            ']' -> throw IllegalSyntaxException("read: unexpected vector terminator: " + c)
            else -> {
                val s = c + readUntilDelimiter()
                /* Read true and false as #t and #f */
                when (s) {
                    "true"  -> return true
                    "false" -> return false
                    else    -> return Symbol.intern(s)
                }
            }
        }
    }

    @Throws(IOException::class)
    private fun readHash(): Any {
        val c = reader.read().toChar()
        if (c == '(') {
            /* Read Quoted vector #(...) */
            val vector = readVector(')')
            if (vector.isEmpty()) {
                return vector
            }
            return Cons.list<AFn>(Quote.QUOTE_SYMBOL, vector)
        } else if (c == '{') {
            return readSet()
        } else if (c == '\\') {
            return readCharacter()
        } else if (c == 't' || c == 'T') {
            return java.lang.Boolean.TRUE
        } else if (c == 'f' || c == 'F') {
            return java.lang.Boolean.FALSE
        } else if (c == '"') {
            return readRegex()
        } else if (isRadix(c) || isExactness(c)) {
            /* Read identifier, not a number */
            val number = "#" + c + readUntilDelimiter()
            /* Read radix and/or exactness and a number */
            var radix: Char? = null
            var exactness: Char? = null
            var restNumber = number
            while (restNumber.length > 1 && restNumber[0] == '#') {
                val ch = restNumber[1]
                if (isExactness(ch)) {
                    if (exactness != null) {
                        throw IllegalSyntaxException("read: bad number: " + number)
                    }
                    exactness = ch
                    restNumber = restNumber.substring(2)
                    continue
                }
                if (isRadix(ch)) {
                    if (radix != null) {
                        throw IllegalSyntaxException("read: bad number: " + number)
                    }
                    radix = ch
                    restNumber = restNumber.substring(2)
                    continue
                }
                break
            }

            if (restNumber.isEmpty() || "+" == restNumber || "-" == restNumber) {
                throw IllegalSyntaxException("read: bad number: " + number)
            }

            /* Check if this is a proper number */
            val result = preProcessNumber(restNumber, exactness, getRadixByChar(radix)) as? Number ?: throw IllegalSyntaxException("read: bad number: " + number)
            return result
        }
        /* Bad hash syntax: read token and throw exception */
        val token = StringBuilder("#")
        if (isValid(c.toInt())) {
            token.append(c)
        }
        if (!Character.isWhitespace(c)) {
            token.append(readUntilDelimiter())
        }
        throw IllegalSyntaxException("read: bad syntax: " + token.toString())
    }

    /**
     * Read a quoted form abbreviation

     * Syntax:
     * <quote>            -> '<form>
     * <quasiquote>       -> `<form>
     * <unquote>          -> ,<form>
     * <unquote-splicing> -> ,@<form>
    </form></unquote-splicing></form></unquote></form></quasiquote></form></quote> */
    @Throws(IOException::class)
    private fun readQuote(c: Char): List<*> {
        var symbol: Symbol? = null
        if (c == '\'') {
            symbol = Quote.QUOTE_SYMBOL
        } else if (c == '`') {
            symbol = Quasiquote.QUASIQUOTE_SYMBOL
        } else if (c == ',') {
            val next = reader.read().toChar()
            if (next == '@') {
                symbol = UnquoteSplicing.UNQUOTE_SPLICING_SYMBOL
            } else {
                reader.unread(next.toInt())
                symbol = Unquote.UNQUOTE_SYMBOL
            }
        }
        return Cons.list<Any>(symbol, nextNonNullToken())
    }

    /**
     * Read a comment

     * Syntax:
     * <comment> --> ;  <all subsequent characters up to a line break>
    </all></comment> */
    @Throws(IOException::class)
    private fun readComment(): String? {
        var i = reader.read()
        while (isValid(i) && !isLineBreak(i.toChar())) {
            /* Read everything until line break */
            i = reader.read()
        }
        /* Comments are ignored, return null */
        return null
    }

    /**
     * Read a String
     * Always returns immutable String

     * Syntax:
     * <string> --> "<string element>*"
     * <string element> --> <any character other than></any>" or \> | \" | \\
    </string></string></string> */
    @Throws(IOException::class)
    private fun readString(): String {
        val string = StringBuilder()
        var i = reader.read()
        var c = i.toChar()
        while (isValid(i) && c != '"') {
            /* Escaping */
            if (c == '\\') {
                val next = reader.read().toChar()
                /* Unicode followed by a hexadecimal number */
                if (next == 'u' || next == 'U') {
                    reader.unread(next.toInt())
                    val chr = readCharacter()
                    if (chr == next) {
                        throw IllegalSyntaxException("read: no hex digit following \\u in string")
                    }
                    string.append(chr)
                } else {
                    /* Check that escape sequence is valid */
                    val character = ESCAPED[next] ?: throw IllegalSyntaxException(String.format("read: unknown escape sequence \\%s in string", next))
                    string.append(character)
                }
            } else {
                string.append(c)
            }
            i = reader.read()
            c = i.toChar()
        }
        /* Always intern Strings read by Reader */
        return string.toString().intern()
    }

    @Throws(IOException::class)
    private fun readRegex(): Pattern {
        val regex = StringBuilder()
        var i = reader.read()
        var c = i.toChar()
        while (isValid(i) && c != '"') {
            regex.append(c)
            if (c == '\\') {
                regex.append(reader.read().toChar())
            }
            i = reader.read()
            c = i.toChar()
        }
        return Pattern.compile(regex.toString())
    }

    /**
     * Read a Character

     * Syntax:
     * <character> --> #\ <any character> | #\ <character name>
     * <character name> --> space | newline
    </character></character></any></character> */
    @Throws(IOException::class)
    private fun readCharacter(): Char {
        val first = reader.read()
        var rest = readUntilDelimiter()
        if (rest.isEmpty()) {
            return first.toChar()
        }
        /* Check if it is a codepoint */
        var radix = 16
        var isCodepoint = first.toChar() == 'u' || first.toChar() == 'U'
        if (Character.isDigit(first.toChar())) {
            radix = 8
            rest = first.toChar() + rest
            isCodepoint = true
        }
        if (!isValidForRadix(rest[0], radix)) {
            isCodepoint = false
        }
        if (isCodepoint) {
            val codepoint = preProcessNumber(rest, 'e', radix) as? Number ?: throw IllegalSyntaxException("read: no hex digit following \\u in string")
            return codepoint.toInt().toChar()
        }
        /* Must be a named char */
        val character = first.toChar() + rest
        if ("linefeed" == character) {
            return NAMED_CHARS["newline"]!!
        }
        val namedChar = NAMED_CHARS[character] ?: throw IllegalSyntaxException("read: bad character constant: #\\" + character)
        return namedChar
    }

    /**
     * Read list

     * Syntax:
     * <list> -> (<list_contents>)
    </list_contents></list> */
    @Throws(IOException::class)
    private fun readList(allowImproperList: Boolean, terminator: Char): Cons<Any> {
        var list: Cons<Any> = Cons.EMPTY
        /* Remember position of a dot (if we meet it) */
        var dotPos = -1
        var i = reader.read()
        var c = i.toChar()
        while (isValid(i) && c != terminator) {
            /* Skip whitespaces */
            while (Character.isWhitespace(c)) {
                c = reader.read().toChar()
            }
            if (c == terminator) {
                break
            }
            reader.unread(c.toInt())
            val token = nextToken()
            /* Check if current token is a dot */
            if (DOT == token) {
                if (!allowImproperList || dotPos > -1) {
                    throw IllegalSyntaxException("read: illegal use of '.'")
                }
                /* Remember the dot position */
                dotPos = list.size
                /* Dot Special Form is allowed as the first element of a list */
                if (dotPos == 0) {
                    list = Cons.list<Any>(DOT)
                }
            } else if (token != null) {
                /* List is empty so far */
                if (list.isEmpty()) {
                    /* Initialize list with the first element (can't modify EMPTY) */
                    list = Cons.list(token)
                } else {
                    /* Add list element */
                    list.add(token)
                }
            }
            i = reader.read()
            c = i.toChar()
        }
        /* Was it a proper list or dot is the first element? */
        if (dotPos < 1) {
            return list
        }
        /* Process improper list */
        if (dotPos != list.size - 1) {
            throw IllegalSyntaxException("read: illegal use of '.'")
        }
        /* Convert list into cons */
        return list.toCons()
    }

    /**
     * Read vector

     * Syntax:
     * <vector> -> #(<vector_contents>)
    </vector_contents></vector> */
    @Throws(IOException::class)
    private fun readVector(terminator: Char): MutableVector {
        /* Improper lists are not allowed */
        return MutableVector(*readList(false, terminator).toTypedArray())
    }

    /**
     * Read hashmap

     * Syntax:
     * <hashmap> -> {<key1> <value1>, ..., <keyN> <valueN>}
    </valueN></keyN></value1></key1></hashmap> */
    @Throws(IOException::class)
    private fun readHashmap(): Map<Any?, Any?> {
        val hashmap= HashMap<Any?, Any?>()
        var i = reader.read()
        var c = i.toChar()
        while (isValid(i) && c != '}') {
            /* Skip whitespaces and commas */
            while (Character.isWhitespace(c) || c == ',') {
                c = reader.read().toChar()
            }
            if (c == '}') break
            reader.unread(c.toInt())
            val key = nextToken()

            /* Skip whitespaces and commas */
            c = reader.read().toChar()
            while (Character.isWhitespace(c) || c == ',') {
                c = reader.read().toChar()
            }
            if (c == '}') break
            reader.unread(c.toInt())
            val value = nextToken()

            hashmap.put(key, value)
            i = reader.read()
            c = i.toChar()
        }
        return hashmap
    }

    /**
     * Read set

     * Syntax:
     * <set> -> #{<value1>, ..., <valueN>}
    </valueN></value1></set> */
    @Throws(IOException::class)
    private fun readSet(): Set<Any?> {
        val set = HashSet<Any?>()
        var i = reader.read()
        var c = i.toChar()
        while (isValid(i) && c != '}') {
            /* Skip whitespaces and commas */
            while (Character.isWhitespace(c)) {
                c = reader.read().toChar()
            }
            if (c == '}') break
            reader.unread(c.toInt())
            set.add(nextToken())
            i = reader.read()
            c = i.toChar()
        }
        return set
    }

    /**
     * Read keyword

     * Syntax:
     * <keyword> -> :<token>
    </token></keyword> */
    @Throws(IOException::class)
    private fun readKeyword(): Keyword {
        val s = readUntilDelimiter()
        if (s.isEmpty()) {
            throw IllegalSyntaxException("read: illegal use of :")
        }
        return Keyword.intern(s)
    }

    /**
     * Deref shortcut

     * \@f -> (deref f)
     */
    @Throws(IOException::class)
    private fun readDeref(): List<Any> {
        return Cons.list(DEREF, nextNonNullToken())
    }

}